llama2.mojo

```diff
@@ -632,27 +649,42 @@ fn transformer(
         # Apply RoPE rotation to the q and k vectors for each head
         let q = state.q.data
         let k = state.k.data
-        for i in range(0, head_size * config.n_kv_heads, 2):
-            let head_dim_half = i % head_size // 2
-            let fcr = freq_cis_real_row.offset(head_dim_half).load(0)
-            let fci = freq_cis_imag_row.offset(head_dim_half).load(0)
-            let q0 = q.offset(i).load(0)
-            let q1 = q.offset(i + 1).load(0)
-            let k0 = k.offset(i).load(0)
-            let k1 = k.offset(i + 1).load(0)
-            q.offset(i).store(0, q0 * fcr - q1 * fci)
-            q.offset(i + 1).store(0, q0 * fci + q1 * fcr)
-            k.offset(i).store(0, k0 * fcr - k1 * fci)
-            k.offset(i + 1).store(0, k0 * fci + k1 * fcr)
-
-        for i in range(head_size * config.n_kv_heads, dim, 2):
-            let head_dim_half = i % head_size // 2
-            let fcr = freq_cis_real_row.offset(head_dim_half).load(0)
-            let fci = freq_cis_imag_row.offset(head_dim_half).load(0)
-            let q0 = q.offset(i).load(0)
-            let q1 = q.offset(i + 1).load(0)
-            q.offset(i).store(0, q0 * fcr - q1 * fci)
-            q.offset(i + 1).store(0, q0 * fci + q1 * fcr)
+
+        # a dirty method to check whether the model is tinyllama-1.1B
```

Comment:
Ugh.. This one looks ugly, probably if we need that much customization, I'd move this to a separate mojo file or github branch.

Reply:
The reason is that **tinyllamas**'s story model's **RoPE** is different from that of  normal llama model (like **llama2, vicuna**).
![image](https://github.com/tairov/llama2.mojo/assets/143765148/f5738e12-f00a-4cc3-8c99-b7953557beeb)

Reply:
So the upper one is the rope in tinyllamas story.
The down is rope in tinyllama or other llama model.

Reply:
@magician-blue is it possible to solve this via pre-converting tinyllama model to llama2?

Reply:
@tairov Possible!

Reply:
To do:
- add new section in readme
- check the implementation ROPE of tinyllama stories in llama2.c, have a look of llama.cpp's rope
- convert tinyllama to normal llama rope and publish to HF
- build a new convertor.py to convert normal llama architecture model (and tokenizer)

Reply:
Yeah, got it. 
I'll think about introducing some packages structure , probably it's worthwhile to separate some parts into modules for supporting different type of models . 
But then this repo won't be anymore Inference in a single file 😀

Reply:
I have checked the ROPE part of llama.cpp.
I found there was 2 type of rope.
- GPT-J style(common):  rotate pairs of even and odd dimensions (llama2, alpcaca, tinyllamas-story)
- GPT-NeoX style: rotate1st half and 2nd half (tinyllama-1.1B, baichuan, falcon)

The design a function rotary(mode).
mode=0 -> common style.
model=1 -> GPT-NeoX style

Reply:
llama.cpp also have model converter https://github.com/ggerganov/llama.cpp/blob/c091cdfb24621710c617ea85c92fcd347d0bf340/convert-gptneox-hf-to-gguf.py#L123

Reply:
> llama.cpp also have model converter https://github.com/ggerganov/llama.cpp/blob/c091cdfb24621710c617ea85c92fcd347d0bf340/convert-gptneox-hf-to-gguf.py#L123

The implementation of GPTneox and that of llama are in different function.